system/datatypes/tagindex.php

```diff
@@ -0,0 +1,94 @@
+<?php
+/*
+	Class: tagindexpage
+	Shows an index page for articles based on their tags
+*/
+	class tagindexpage extends HyphaSystemPage {
+		function __construct(RequestContext $O_O) {
+			parent::__construct($O_O);
+		}
+
+		public function process(HyphaRequest $request) {
+			return $this->tagIndexView($request);
+		}
+
+		public function tagIndexView(HyphaRequest $request) {
+			$lang = $request->getArg(0);
+			// TODO: This urldecode should really be done by
+			// HyphaRequest (#274)
+			$label = urldecode($request->getArg(1));
+
+			$tag = HyphaTags::findTagByLabel($lang, $label);
+			if ($tag === null)
+				return '404';
+
+			/** @var HyphaDomElement $main */
+			$main = $this->html->find('#main');
+
+			/** @var HyphaDomElement $pagename */
+			$pagename = $this->html->find('#pagename');
+			$tagprefixspan = $this->html->createElement('span');
+			$tagprefixspan->setText('Index for tag: ');
+			$tagprefixspan->setAttribute('class', 'tag_prefix');
+			$tagspan = $this->html->createElement('span');
+			$tagspan->setAttribute('class', 'selected_tag');
+			$tagspan->setText($label);
+			$pagename->append($tagprefixspan);
+			$pagename->append($tagspan);
+
+			$includePrivate = isUser();
+			$pages = $this->getPages($tag, $includePrivate);
+			$pages = $this->sortPages($pages);
+
+			$pages_list = $this->renderPages($pages);
+			$main->append($pages_list);
+		}
+
+		private function sortPages($pages) {
+			// TODO: This should cache sort dates
+			$compare = function($a, $b) {
+				$a_date = $a->getSortDateTime();
+				$b_date = $b->getSortDateTime();
+				// TODO Can we better handle null here?
+				return ($b_date ? $b_date->getTimestamp() : 0) - ($a_date ? $a_date->getTimestamp() : 0);
+			};
+			usort($pages, $compare);
+			return $pages;
+		}
+
+		private function getPages(HyphaTag $tag, $includePrivate = false) {
+			$pagetypes = ['peer_reviewed_article'];
+			$pagetypes = [];
+			$pagenodes = HyphaTags::findPagesWithTags($tag, $pagetypes, $includePrivate);
+			$pages = [];
+			foreach ($pagenodes as $node) {
+				$pages[] = createPageInstance($this->O_O, $node);
+			}
+			return $pages;
+		}
+
+		private function renderPages($pages) {
+			/** @var HyphaDomElement $ul */
+			$ul = $this->html->createElement('ul');
+			$ul->addClass('tagindex');
+			foreach ($pages as $page) {
+				$li = $this->renderPage($page);
+				$ul->append($li);
+			}
+			return $ul;
+		}
+
+		private function renderPage(HyphaDatatypePage $page) {
+			$li = $this->html->createElement('li');
+			$li->addClass('tagindex-item');
+			$li->addClass('type_' . get_class($page));
+			$li->addClass($page->privateFlag ? 'is-private' : 'is-public');
+
+			$a = $this->html->createElement('a')->appendTo($li);
+			$a->setAttribute('href', $page->language.'/'.$page->pagename);
```

Comment:
are special characters handled here? (double checking because I read something in the chat)

Reply:
This leaves urlencoding of special characters (in particular spaces) to the browser, at some point url generation should really be generalized and do urlencoding explicitly. Special characters are handled at the HTML level properly by DOMDocument here.